Jetpack Boost: Strict TypeScript

pyronaur opened the pull request:
Guardrails off.

IIRC we enabled typescript without strict mode a long time ago because we were "ramping up slowly". I think 2 years is slow enough 😛 

On top of that, this addresses missing React component types.

## Proposed changes:
Enables strict mode in TypeScript.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:

Apply patch, create a function with unspecified types, see squiggly lines:

![2023_12_13RaalHoSH 42](https://github.com/Automattic/jetpack/assets/988095/61f88b43-d718-4451-b9e3-5e7e6080f926)

pyronaur committed:
Ensure that TypeScript checks react types

pyronaur committed:
Changelog

thingalon reviewed (dismissed):
Try running `pnpm run build-development` and you will see a giant list of errors. E2E tests and Github build fail too.

While I agree with this idea, I'm not willing to put trunk in a state where it can't be compiled.

pyronaur committed:
Production build compiles

pyronaur committed:
Importing jsx is pointless

pyronaur committed:
Clean-up dev-scripts

pyronaur committed:
Disable strict mode for compile-ts

pyronaur committed:
Enable strict mode

pyronaur committed:
Use NODE_ENV=production

Instead of relying on ROLLUP_WATCH to detect a development environment - look for NODE_ENV=production to determine what's a development build.

pyronaur committed:
Disable strict mode in rollup-dev

pyronaur committed:
Tweak development & production build settings

And add a comment about strict mode.

Right now - we need to disable strict mode in development because it's too noisy - but eventually we can either fully enable strict mode, or just for development.

pyronaur commented:
@thingalon The project now compiles.

1 - Strict mode is enabled in tsconfig - IDE's should pick up typescript files strictly
2 - Rollup doesn't actually validate TypeScript - it just compiles both in development and production. This now remains unchanged
3 - Updated the dev workflow a bit too to also watch for non-strict TypeScript errors during development by default.
4 - Added commands `pnpm run dev-compile` and `pnpm run dev-validate-ts`. They're both run concurrently in dev mode (default watch for jetpack cli), but you can run them separately in two panes in iTerm :)

haqadn reviewed (approved):
🥇
